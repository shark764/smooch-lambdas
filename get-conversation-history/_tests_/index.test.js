const axios = require('axios');

jest.mock('aws-sdk');
jest.mock('axios');

const event = {
  params: {
    'tenant-id': '66d83870-30df-4a3b-8801-59edff162034',
    'interaction-id': '667802d8-2260-436c-958a-2ee0f71f73f0',
  },
};

const mockGetSecretValue = jest.fn(() => {})
  .mockImplementation(() => ({
    promise: () => ({
      SecretString: JSON.stringify({
        '5e31c81640a22c000f5d7f28-id': 'id',
        '5e31c81640a22c000f5d7f28-secret': 'secret',
      }),
    }),
  }));

const mockGetMessages = jest.fn()
  .mockImplementation(() => ({
    messages: [
      {
        type: '',
        _id: '',
        role: '',
        metadata: {
          resourceId: '',
        },
        from: '',
        mediaUrl: 'url1',
        mediaType: 'txt',
        mediaSize: '100kb',
        received: '50',
      },
    ],
  }));

axios.mockImplementation(() => ({
  data: {
    appId: '5e31c81640a22c000f5d7f28',
    userId: '667802d8-2260-436c-958a-2ee0f71f73f1',
    customer: 'mock-customer',
  },
}));

const mockSmoochCore = jest.fn().mockImplementation(() => ({
  appUsers: {
    getMessages: mockGetMessages,
  },
}));

jest.mock('aws-sdk', () => ({
  SecretsManager: jest.fn().mockImplementation(() => ({ getSecretValue: mockGetSecretValue })),
}));


jest.mock('smooch-core', () => mockSmoochCore);

const { handler } = require('../index');

describe('get-conversation-history', () => {
  describe('Everthing is successful', () => {
    it("messages are filtered for type 'formResponse' and quoted messages and are also mapped for type 'formResponse'", async () => {
      mockGetMessages.mockImplementationOnce(() => ({
        messages: [
          {
            type: 'formResponse',
            quotedMessage: {
              content: {
                metadata: 'meta-data',
              },
            },
            fields: [{ text: 'collect-message response' }],
            textFallback: 'collect-message response',
          },
        ],
      }));
      const result = await handler(event);
      expect(result).toMatchSnapshot();
    });

    it("messages are filtered for role 'appUser' and type not equal to 'formResponse'", async () => {
      mockGetMessages.mockImplementationOnce(() => ({
        messages: [
          {
            type: '',
            role: 'appUser',
          },
        ],
      }));
      const result = await handler(event);
      expect(result).toMatchSnapshot();
    });

    it('messages are filtered for metadata', async () => {
      mockGetMessages.mockImplementationOnce(() => ({
        messages: [
          {
            metadata: {
              resourceId: '66d83870-30df-4a3a-8801-59edff162037',
              from: '',
            },
          },
        ],
      }));
      const result = await handler(event);
      expect(result).toMatchSnapshot();
    });

    it('when no filter is applied', async () => {
      mockGetMessages.mockImplementationOnce(() => ({
        messages: [
          {
            role: '',
            type: '',
          },
        ],
      }));
      const result = await handler(event);
      expect(result).toMatchSnapshot();
    });

    it("messages are mapped for the role 'appMaker'", async () => {
      mockGetMessages.mockImplementationOnce(() => ({
        messages: [
          {
            role: 'appMaker',
            metadata: {
              type: 'agent',
              from: '',
            },
          },
        ],
      }));
      const result = await handler(event);
      expect(result).toMatchSnapshot();
    });

    it("messages are mapped for the role not equal to 'appMaker'", async () => {
      mockGetMessages.mockImplementationOnce(() => ({
        messages: [
          {
            role: '',
            metadata: {},
            text: 'normal messages',
          },
        ],
      }));
      const result = await handler(event);
      expect(result).toMatchSnapshot();
    });

    it("messages are mapped for the role 'appMaker' and type 'form'", async () => {
      mockGetMessages.mockImplementationOnce(() => ({
        messages: [
          {
            role: 'appMaker',
            type: 'form',
            metadata: {
              type: 'agent',
              from: '',
            },
            fields: [{ label: 'collect-message', name: 'collect-message' }],
          },
        ],
      }));
      const result = await handler(event);
      expect(result).toMatchSnapshot();
    });
    describe('Walkthrough', () => {
      beforeAll(async () => {
        await handler(event);
      });
      it('passes in the correct arguments to secretsClient.getSecretValue()', async () => {
        expect(mockGetSecretValue.mock.calls).toMatchSnapshot();
      });

      it('passes in the correct arguments to smooch.appUsers.getMessages()', async () => {
        expect(mockGetMessages.mock.calls).toMatchSnapshot();
      });
    });
  });
  it('sends back status 500 when there is a problem fetching interaction messages', async () => {
    mockGetMessages.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 if there is an error retrieving digital channels credentials (error by SmoochCore)', async () => {
    mockSmoochCore.mockImplementationOnce(() => {
      throw new Error('SmoochCore');
    });
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 if there is an error retrieving cx credentials', async () => {
    mockGetSecretValue.mockImplementationOnce(() => ({
      promise: () => ({
        SecretString: JSON.stringify({
          '5e31c81640a22c000f5d7f28-id': 'id',
          '5e31c81640a22c000f5d7f28-secret': 'secret',
        }),
      }),
    }));
    mockGetSecretValue.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is an error in retrieving the interaction metadata', async () => {
    axios.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is an error retrieving digital channels credentials (error by getSecretValue())', async () => {
    mockGetSecretValue.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });
});
