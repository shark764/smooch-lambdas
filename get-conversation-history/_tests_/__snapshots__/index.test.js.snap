// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-conversation-history Everthing is successful Walkthrough passes in the correct arguments to secretsClient.getSecretValue() 1`] = `
Array [
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-app",
    },
  ],
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-cx",
    },
  ],
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-app",
    },
  ],
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-cx",
    },
  ],
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-app",
    },
  ],
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-cx",
    },
  ],
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-app",
    },
  ],
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-cx",
    },
  ],
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-app",
    },
  ],
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-cx",
    },
  ],
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-app",
    },
  ],
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-cx",
    },
  ],
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-app",
    },
  ],
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-cx",
    },
  ],
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-app",
    },
  ],
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-cx",
    },
  ],
]
`;

exports[`get-conversation-history Everthing is successful Walkthrough passes in the correct arguments to smooch.appUsers.getMessages() 1`] = `
Array [
  Array [
    Object {
      "appId": "5e31c81640a22c000f5d7f28",
      "userId": "667802d8-2260-436c-958a-2ee0f71f73f1",
    },
  ],
  Array [
    Object {
      "appId": "5e31c81640a22c000f5d7f28",
      "userId": "667802d8-2260-436c-958a-2ee0f71f73f1",
    },
  ],
  Array [
    Object {
      "appId": "5e31c81640a22c000f5d7f28",
      "userId": "667802d8-2260-436c-958a-2ee0f71f73f1",
    },
  ],
  Array [
    Object {
      "appId": "5e31c81640a22c000f5d7f28",
      "userId": "667802d8-2260-436c-958a-2ee0f71f73f1",
    },
  ],
  Array [
    Object {
      "appId": "5e31c81640a22c000f5d7f28",
      "userId": "667802d8-2260-436c-958a-2ee0f71f73f1",
    },
  ],
  Array [
    Object {
      "appId": "5e31c81640a22c000f5d7f28",
      "userId": "667802d8-2260-436c-958a-2ee0f71f73f1",
    },
  ],
  Array [
    Object {
      "appId": "5e31c81640a22c000f5d7f28",
      "userId": "667802d8-2260-436c-958a-2ee0f71f73f1",
    },
  ],
  Array [
    Object {
      "appId": "5e31c81640a22c000f5d7f28",
      "userId": "667802d8-2260-436c-958a-2ee0f71f73f1",
    },
  ],
]
`;

exports[`get-conversation-history Everthing is successful messages are filtered for metadata 1`] = `
Object {
  "body": Object {
    "interactionId": "667802d8-2260-436c-958a-2ee0f71f73f0",
    "messages": Array [
      Object {
        "contentType": undefined,
        "file": Object {
          "mediaSize": undefined,
          "mediaType": undefined,
          "mediaUrl": undefined,
        },
        "from": "mock-customer",
        "id": undefined,
        "quotedMessage": Object {},
        "resourceId": null,
        "text": undefined,
        "timestamp": NaN,
        "type": "customer",
      },
    ],
  },
  "status": 200,
}
`;

exports[`get-conversation-history Everthing is successful messages are filtered for role 'appUser' and type not equal to 'formResponse' 1`] = `
Object {
  "body": Object {
    "interactionId": "667802d8-2260-436c-958a-2ee0f71f73f0",
    "messages": Array [
      Object {
        "contentType": "",
        "file": Object {
          "mediaSize": undefined,
          "mediaType": undefined,
          "mediaUrl": undefined,
        },
        "from": "mock-customer",
        "id": undefined,
        "quotedMessage": Object {},
        "resourceId": null,
        "text": undefined,
        "timestamp": NaN,
        "type": "customer",
      },
    ],
  },
  "status": 200,
}
`;

exports[`get-conversation-history Everthing is successful messages are filtered for type 'formResponse' and quoted messages and are also mapped for type 'formResponse' 1`] = `
Object {
  "body": Object {
    "interactionId": "667802d8-2260-436c-958a-2ee0f71f73f0",
    "messages": Array [
      Object {
        "contentType": "formResponse",
        "file": Object {
          "mediaSize": undefined,
          "mediaType": undefined,
          "mediaUrl": undefined,
        },
        "from": "mock-customer",
        "id": undefined,
        "quotedMessage": Object {
          "content": Object {
            "file": Object {},
            "id": undefined,
            "text": undefined,
            "type": undefined,
          },
        },
        "resourceId": null,
        "text": "collect-message response",
        "timestamp": NaN,
        "type": "customer",
      },
    ],
  },
  "status": 200,
}
`;

exports[`get-conversation-history Everthing is successful messages are mapped for the role 'appMaker' 1`] = `
Object {
  "body": Object {
    "interactionId": "667802d8-2260-436c-958a-2ee0f71f73f0",
    "messages": Array [
      Object {
        "contentType": undefined,
        "file": Object {
          "mediaSize": undefined,
          "mediaType": undefined,
          "mediaUrl": undefined,
        },
        "from": "",
        "id": undefined,
        "quotedMessage": Object {},
        "resourceId": undefined,
        "text": undefined,
        "timestamp": NaN,
        "type": "agent",
      },
    ],
  },
  "status": 200,
}
`;

exports[`get-conversation-history Everthing is successful messages are mapped for the role 'appMaker' and type 'form' 1`] = `
Object {
  "body": Object {
    "interactionId": "667802d8-2260-436c-958a-2ee0f71f73f0",
    "messages": Array [
      Object {
        "contentType": "form",
        "file": Object {
          "mediaSize": undefined,
          "mediaType": undefined,
          "mediaUrl": undefined,
        },
        "from": "",
        "id": undefined,
        "quotedMessage": Object {},
        "resourceId": undefined,
        "text": "collect-message",
        "timestamp": NaN,
        "type": "agent",
      },
    ],
  },
  "status": 200,
}
`;

exports[`get-conversation-history Everthing is successful messages are mapped for the role not equal to 'appMaker' 1`] = `
Object {
  "body": Object {
    "interactionId": "667802d8-2260-436c-958a-2ee0f71f73f0",
    "messages": Array [
      Object {
        "contentType": undefined,
        "file": Object {
          "mediaSize": undefined,
          "mediaType": undefined,
          "mediaUrl": undefined,
        },
        "from": "mock-customer",
        "id": undefined,
        "quotedMessage": Object {},
        "resourceId": null,
        "text": "normal messages",
        "timestamp": NaN,
        "type": "customer",
      },
    ],
  },
  "status": 200,
}
`;

exports[`get-conversation-history Everthing is successful when no filter is applied 1`] = `
Object {
  "body": Object {
    "interactionId": "667802d8-2260-436c-958a-2ee0f71f73f0",
    "messages": Array [],
  },
  "status": 200,
}
`;

exports[`get-conversation-history sends back status 500 if there is an error retrieving cx credentials 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to retrieve cx credentials",
  },
  "status": 500,
}
`;

exports[`get-conversation-history sends back status 500 if there is an error retrieving digital channels credentials (error by SmoochCore) 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to retrieve digital channels credentials",
  },
  "status": 500,
}
`;

exports[`get-conversation-history sends back status 500 when there is a problem fetching interaction messages 1`] = `
Object {
  "body": Object {
    "message": "An error occurred fetching interaction messages",
  },
  "status": 500,
}
`;

exports[`get-conversation-history sends back status 500 when there is an error in retrieving the interaction metadata 1`] = `
Object {
  "body": Object {
    "message": "An error occurred retrieving the interaction metadata",
  },
  "status": 500,
}
`;

exports[`get-conversation-history sends back status 500 when there is an error retrieving digital channels credentials (error by getSecretValue()) 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to retrieve digital channels credentials",
  },
  "status": 500,
}
`;
